active-rfcs/0000-function-api.md

```diff
@@ -0,0 +1,774 @@
+
+ ``` js
+// read the value
+console.log(msg.value) // 'hello'
+ // mutate the value
+msg.value = 'bye'
+```
+
+### Why do we need value wrappers?
+
+Primitive values in JavaScript like numbers and strings are not passed by reference. Returning a primitive value from a function means the receiving function will not be able to read the latest value when the original is mutated or replaced.
+
+Value wrappers are important because they provide a way to pass around mutable and reactive references for arbitrary value types. This is what enables composition functions to encapsulate the logic that manages the state while passing the state back to the components as a trackable reference:
+
+``` js
+setup() {
+  const valueA = useLogicA() // logic inside useLogicA may mutate valueA
+  const valueB = useLogicB()
+  return {
+    valueA,
+    valueB
+  }
+}
+```
+
+Value wrappers can also hold non-primitive values and will make all nested properties reactive. Holding non-primitive values like objects and arrays inside a value wrapper provides the ability to entirely replace the value with a fresh one:
+
+``` js
+const numbers = value([1, 2, 3])
+// replace the array with a filtered copy
+numbers.value = numbers.value.filter(n => n > 1)
+```
+
+If you want to create a non-wrapped reactive object, use `state` (which is an exact equivalent of 2.x `Vue.observable` API):
+
+``` js
+import { state } from 'vue'
+
+const object = state({
+  count: 0
+})
+
+object.count++
+```
+
+### Value Unwrapping
+
+Note in the last example we are using `{{ msg }}` in the template without the `.value` property access. This is because **value wrappers get "unwrapped" when they are accessed on the render context or as a nested property inside a reactive object.**
+
+You can mutate an unwrapped value binding in inline handlers:
+
+``` js
+const MyComponent = {
+  setup() {
+    return {
+      count: value(0)
+    }
+  },
+  template: `<button @click="count++">{{ count }}</button>`
+}
+```
+
+Value wrappers are also automatically unwrapped when accessed as a nested property inside a reactive object:
+
+``` js
+const count = value(0)
+const obj = state({
+  count
+})
+
+console.log(obj.count) // 0
+
+obj.count++
+console.log(obj.count) // 1
+console.log(count.value) // 1
+
+count.value++
+console.log(obj.count) // 2
+console.log(count.value) // 2
+```
+
+As a rule of thumb, the only occasions where you need to use `.value` is when directly accessing value wrappers as variables.
+
+### Bindings in Manual Render Functions
+
+Bindings returned by `setup()` is also usable inside manually written render functions - they are exposed on `this`. In addition, the render function receives the internal component instance as the argument, from which `state`, `props` and `slots` can be destructured:
+
+``` js
+const MyComponent = {
+  props: {
+    msg: String
+  },
+  setup() {
+    return {
+      count: value(0)
+    }
+  },
+  render({ state, props, slots }) {
```

Comment:
I think it's not quite clear how to use this `render` function.
- Are we getting `createElement` from the params? Are we expected to use that?
- Is this method supposed to return `VNode`?

I think it would be nice to compete this example and show how you would render something (`h1`?) that contains values from your state/props (like `msg` and/or `count`)

Reply:
https://github.com/vuejs/rfcs/pull/28